### DISCUSSION TITLE
Consider removing ISerializable support

### DISCUSSION THREAD

**Discussion by bradwilson** [xunit/xunit#2506] [Feedback Request]
## Update: July 6, 2022

_The solution I landed on includes removing the type-unsafe serialization format in favor of a mostly lookup table-based type system, while still supporting serialization of enums and `IXunitSerializable` objects with embedded type information. This limits the type-embedded serialization to just those two cases, and we verify that we are only creating enums or classes that implement `IXunitSerializable` when deserializing. Since this interface was designed explicitly only for unit testing with xUnit.net (and it's not an arbitrary serialization interface like `ISerializable`), this should alleviate any potential security concerns. In practice, this should not noticeably change the kinds of data that v3 can serialize when compared to v2, but it should be done in a much safer manner (and without relying on any unsafe built-in serialization infrastructure), and it should be both more memory efficient and higher performance. You can [read more about this below](#discussioncomment-3089302)._

## Description of the issue

Currently we use [`BinaryFormatter`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.formatters.binary.binaryformatter) to support [serializing](https://github.com/xunit/xunit/blob/c27a91f8cbcee37cb45699a3a81287bca225e876/src/xunit.v3.common/Utility/SerializationHelper.cs) arbitrary test case data across process boundaries. This replaces support for [`IXunitSerializable`](https://github.com/xunit/abstractions.xunit/blob/main/src/xunit.abstractions/Serialization/IXunitSerializable.cs) from v2 (which was added because the lowest common denominator of target frameworks for v2 does not include `ISerializable`). This serializer is used today to allow arbitrary data to be passed to test cases via theory rows, and allow the end user to run those individual theory rows in Visual Studio's Test Explorer. Without some kind of serialization support, we would end up reverting back to xUnit.net v1 behavior (one Test Explorer item per test method, which runs all the associated data rows without the ability to run individual rows independently of one another).

Binary serialization has been [deprecated because of security concerns](https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide) and will be outright removed from future versions of .NET. Other related serializers (SoapFormatter, LosFormatter, NetDataContractSerializer, ObjectStateFormatter) suffer from the same concerns. A quick analysis of the serialization in v2 shows that it, too, suffers from the arbitrary type deserialization concern.

Generally speaking, we have two alternatives.

### 1. Remove generalized object serialization

If we were to remove generalized object serialization, that would mean that any test method which contains unsupported objects in a theory row would revert to the v1 behavior in Test Explorer (this is also what happens today in v2 for arbitrary objects which do not implement `IXunitSerializable`).

What types would be supported in box? Likely the list would look very much like [what we support intrinsically in v2](https://github.com/xunit/xunit/blob/19618efd329b54933d3a8a62933b0121319084a7/src/common/XunitSerializationInfo.cs#L112-L184). Any data row which contains data outside this supported type list would still be supported, but would _lose the ability in Test Explorer to run individual data rows_.

### 2. Add support for a safer serialization alternative

There are some safer alternatives (such as [`XmlSerializer`](https://docs.microsoft.com/en-us/dotnet/api/system.xml.serialization.xmlserializer)) which would allow test authors to implement serialization that would be supported, but an important caveat would be that these serialization formats do not support polymorphism.

What does that mean? In practical terms: if the object in your data row is of type `Employee`, then your test method argument must also be of type `Employee`: trying to use base types (like `Person`) or interfaces (like `IPerson`) would not be supported. During the discovery process, we would have to look for exact type matches between data row element and test method signature, and fall back to the non-serialized behavior when they don't match.

Additionally, there is no good _single_ serialization system to support, since every one of these safer serializers tends to come with its own set of features, limitations, and contracts for how serialization is defined (most tend to use attributes). These feature sets and contracts are not cross-serializer. So it's very likely that we would have to pick just a single (or _very_ small number) of these serializers to support, as we would not be equipped to support an arbitrary number of them.

We could also re-introduce support for something like `IXunitSerializable` but with the new exact-type-match requirements.

## Thoughts

Please provide thoughts on this in this discussion thread. Should we support any kind of arbitrary object serialization? If so, would you prefer something built-in like `XmlSerializer` (and which one?), or would you prefer us to keep `IXunitSerializable` but with the new same-type restrictions?

Thanks!

**Comment by Tornhoof** [comment]
I personally think that the intrinsic types are enough. I can't remember ever looking or needing more than that. So from my POV, your first alternative would be fine.

**Reply by bradwilson** [reply]
Thanks! Even if you have the occasion to use custom data types (and we choose not to support customized serialization), you'll still be able to run those tests...just not individually from Test Explorer (only as a group, per test method instead of per data row).

**Reply by jzabroski** [reply]
Does "intrinsic type" cover types that are "platform intrinsic" but not supported by .NET Attributes due to assembly metadata format limitations, like `decimal` and `DateTime`?

Immo asked a somewhat related question a few months ago, which is, should .NET Attributes allow more types to be stored? https://github.com/dotnet/runtime/issues/4525#issuecomment-1004450575

**Reply by bradwilson** [reply]
This is the current list, as linked above: https://github.com/xunit/xunit/blob/19618efd329b54933d3a8a62933b0121319084a7/src/common/XunitSerializationInfo.cs#L112-L184

I have some work that I started locally (not pushed anywhere yet). Support for `Type` is not likely to survive, but the rest of that list should be supported.

**Reply by jzabroski** [reply]
`Type` seems redundant with open generic support in theories, as my linked example demonstrates. As long as that is covered in regression, I see no need for Type, as I can f-bound `T` in `public void Test<T>`

One hairbrained thought I had is if there would be any benefit to a SerDe approach, similar to @tyrrz Fluent Generics in C#: https://tyrrrz.me/blog/fluent-generics - in other words, each `if` is a closed generic type that can be extended

**Reply by bradwilson** [reply]
I'm not sure yet, but it's possible (likely?) that generics won't be supported for serialization, because it is sort of the definition of polymorphism (and if I can't serialize `Type`, then I can't serialize the discovered generic type(s), and now we're back where we started with disallowing polymorphism).

**Reply by jzabroski** [reply]
I would limit support to public members and use SerDe approach. This view looks at test data as _abstract data types_, rather than _objects_. You serialize _members_ by letting people specify how a member gets de-serialized. Requires public setters/opinionated object model.

**Reply by bradwilson** [reply]
The actual mechanism of serialization is less the issue than the polymorphism is.

Here's an example. Let's say this is your test:

```csharp
[Theory]
[InlineData(24)]
[InlineData("Hello")]
public void MyTest(object data) { ... }
```

This is not legal for serialization any more, and it has nothing to do with the serialization mechanism of the data (we directly support `int` and `string`) but rather that the type of the data does not exactly match the type of the argument. If I have serialized `24` or `"Hello"`, I don't know at deserialization time what the type is, and my only hint it the argument type.

In v2, this is legal, but it runs afoul of the security concerns that caused .NET to basically kill most of the useful forms of `ISerializable`. What we did (and what they did) was store the actual type along side the serialized value so we knew what type to deserialize. The security issue is that it's possible then to construct serialized data that would allow you to deserialize arbitrary things. The only solution to this is to get the type information from _somewhere else_ other than the serialization, and the only other _somewhere else_ available to us is the argument type.

So this is legally serializable:

```csharp
[Theory]
[InlineData(24)]
[InlineData(2112)]
public void MyTest(int data) { ...}
```

This is not:

```csharp
[Theory]
[InlineData(24)]
[InlineData(2112)]
public void MyTest(object data) { ...}
```

As for generics, if I were to write this:

```csharp
[Theory]
[InlineData(24)]
[InlineData(2112)]
public void MyTest<T>(T data) { ...}
```

It's unclear to me how I can make this work without violating the "type should not be part of the serialization" rule, because you need to know `T`, and you can't infer it from the deserialization of the value (nor can you include it in the serialization of the test case in any other way, because then it becomes tamperable).

**Note: This doesn't make any of these tests illegal or incorrect or broken, but they will not be able to be seen as individual data rows inside Visual Studio.**

**Reply by jzabroski** [reply]
> Note: This doesn't make any of these tests illegal or incorrect or broken, but they will not be able to be seen as individual data rows inside Visual Studio.

One way I use theories is [Combinatoric](https://www.nuget.org/packages/Combinatorics/) Input tests / Orthogonal Array Testing Strategy-style tests, where I test a "feature vector" composed of many features. It would be quite painful to need to re-execute the entire orthogonal array because an edge case was detected and failed.

Ultimately, xUnit.net is still the king of .NET test runners / testing frameworks. But, this feature is one of those features I think of as "not your grandpa's test runner" (cough)nunit(cough) that typically come from the [SUnit family tree](https://docs.intersystems.com/irislatest/csp/docbook/DocBook.UI.Page.cls?KEY=TUNT_TESTINGFRAMEWORKS)

For a brief while, MbUnit was cool, and the Gallio Automation Framework actually allowed the FLOSS community to innovate _past Visual Studio Test Explorer_ (which was tightly coupled to MsTest, which everyone can agree was horribly like the original SUnit and very limited in its expressiveness). If you look back at the MbUnit design goals, design tenant number 1 was actually solving this problem we're discussing right now: https://github.com/Gallio/mbunit-v3/blob/master/docs/Gallio%20Ideas.txt

> It's unclear to me how I can make this work without violating the "type should not be part of the serialization" rule, because you need to know `T`, and you can't infer it from the deserialization of the value (nor can you include it in the serialization of the test case in any other way, because then it becomes tamperable).

I was thinking you would use tuples - but your example is using a string to represent the triple - I don't think I'm fully groking the limitation here.  I understand the [ysoserial.net](https://github.com/pwntester/ysoserial.net) type confusion attacks and why open-ended deserialization is very bad and should be banned.  I think the problem needs to be broken down into smaller steps.

Why cant a generic deserializer "pattern match" on an _actual tuple_ (not some weird string encoding) using exact type matching?  That's where I was thinking encoding `if` statements as closed generic types `SerDe<T>` and making a SerDe interface that knows how to SerDe anything that implements `SerDe<T>`.  Then it's really just a value factory/byte factory.

> violating the "type should not be part of the serialization" rule

I'm not clear what that rule is?  I thought the real issue is that deserialization doesn't call a constructor, but rather writes directly to fields via reflection, and that "gadgets" can be built using things like collections (dictionaries) to create "gadget chains" which allow chaining together multiple steps to execute an exploit.

In the Xunit.net case, you're not caching these serialized values anywhere - there is no data-at-rest, it's data-in-transit through IPC between the test runner and the test host (driver program that executes one test in isolation).  Or, is the concern that the `Xunit.Sdk.DataAttribute` interface can be used to load data at rest into some type, and thus cause a type confusion attack?  So, the issue has less to do with passing values from the test runner to the test host, and more to do with the basic mechanism of just serializing the data.  But even so, that would mean we just get rid of stuff that allows directly loading data-at-rest, right?  Xunit hasn't even supported any official external mechanisms in forever - I cant even remember when the ExcelDataAttribute was taken out, but that must've been a decade ago.

**Reply by jzabroski** [reply]
To better understand my thinking, here is the Java solution to the original ysoserial attack in Java by Lawrence and Frohoff.  https://commons.apache.org/proper/commons-io/apidocs/index.html

> `public class` [`ValidatingObjectInputStream`](https://commons.apache.org/proper/commons-io/apidocs/src-html/org/apache/commons/io/serialization/ValidatingObjectInputStream.html#line.45)
> `extends` [`ObjectInputStream`](https://docs.oracle.com/javase/7/docs/api/java/io/ObjectInputStream.html?is-external=true)
> 
> An ObjectInputStream that's restricted to deserialize a limited set of classes.
> Various accept/reject methods allow for specifying which classes can be deserialized.
> Design inspired by [IBM DeveloperWorks Article](http://www.ibm.com/developerworks/library/se-lookahead/).

I'm just proposing you lift the accept/reject method to closed generic types and inform the xunit test runner what those accepts are.  Anything not explicitly an accept is a reject. Any reject is delegated to the next available closed generic type to handle.  (Conceptually - optimizations can be done so it's just a jump table).

**Reply by jzabroski** [reply]
Also, re-reading Strategy 2,

> We could also re-introduce support for something like IXunitSerializable but with the new exact-type-match requirements.

Is this part of Strategy 2, or is this an orthogonal and-also Strategy 3 that could compliment Strategy 1 or 2?

Because, I think this is what I am advocating for.

**Reply by jzabroski** [reply]
Here is the archive.org version of the IBM Dev Works article linked above: https://web.archive.org/web/20160313110353/http://www.ibm.com/developerworks/library/se-lookahead/

**Reply by bradwilson** [reply]
>> violating the "type should not be part of the serialization" rule
>
> I'm not clear what that rule is?

This is the fundamental security problem.

If your serialization includes both the _data_ to be deserialized and the _type_ to deserialize it into, it's vulnerable to attack because the attacker can choose any arbitrary type and data and force you to create it and shove that data into it.

So the serialization **cannot** include the type to be deserialized. The type must be known and fixed. In the case of theory data, the only way to **know** the type without serializing it is if the argument _type_ exactly matches the argument _instance_. This is what I mean by "no polymorphism". The code cannot look at an argument type of `object` and then have a serialized value of type `string` and recreate it, because it has _no way to know that what it has serialized was of type `string`_. If it knew that, then it is vulnerable to attack. That's why generic argument types don't work, either: what is the type of `T`? Generics are by definition polymorphism.

**Reply by bradwilson** [reply]
To be clear: We can solve this for all the built-in types because we can use a serialized "type" that isn't arbitrary (say we use an enum/int value, like 0 = int, 1 = string, etc.). What you can't safely solve is generics/polymorphism for _arbitary_ types.

**Reply by bradwilson** [reply]
If we do `IXunitSerializable`, we would obviously only support deserializing types which implemented it. That would still be arbitrary but at least limited to types known to be "safely deserializable". There would be none of these built into the system so it's hard to imagine an attack that would target anything that wasn't already designed to be recreated in a test environment.

I'll give some thought as to whether this is a reasonable half-measure or not.

**Reply by jzabroski** [reply]
> If your serialization includes both the _data_ to be deserialized and the _type_ to deserialize it into, it's vulnerable to attack because the attacker can choose any arbitrary type and data and force you to create it and shove that data into it.

Thanks, that's really clear.  I think what I was suggesting was that the test library itself do the de-serialization, and so the only way for the type confusion attack to occur would be if a hacker were to somehow inject a SerDe<T> at the class-level of what you are testing.

Put another way... to use your example:

> To be clear: We can solve this for all the built-in types because we can use a serialized "type" that isn't arbitrary (say we use an enum/int value, like 0 = int, 1 = string, etc.).

What you're describing is a **tagged interpreter**. Every **node type** in the **AST** has a **tag** associated with the **node value**, and an **interpreter** traverses the **AST** in order to execute a computation.

What I am describing, is possibly an extension of that solution, but you use closed generic types to define the initial set of tags, and allow consumers to create their own _incremental interpreters_ for other tags not part of the initial set. Ultimately, the T has to be serialized by the test explorer and then deserialized by the test host that executes the test, 

In some ways, this is similar to how OData can only execute a query against a table that is mapped in Entity Framework.  If a database has a salaryInfo table and its not mapped, you aren't getting access to it. The only difference here is the initial set of table tags in OData is empty.

**Reply by jzabroski** [reply]
To further this explanation, Rust's SerDe is not vulnerable in general to the type exploitations you see discussed in ysoserial.net, because the implementer of Deserialize has to pass in a Type to deserialize into. The magic trick is in how to build a safe Serialize Implementation through composition. So, for example, the yaml-rust Crate (aka nuget package in C#) is not vulnerable to ysoserial attacks, but there [many languages with yaml serialization libraries that are vulnerable](https://blogs.embarcadero.com/yaml-and-remote-code-execution).

I think I'm either 100% right or off by a country mile and this will age poorly. :-)  The nuance is in how "arbitrary type" is defined and whether my plug-in model for building serializers actually allows an exploit - I dont think it does.

**Reply by jzabroski** [reply]
Here is an alternative explanation: https://serde.rs/

> Serde is a framework for **ser**ializing and **de**serializing Rust data structures efficiently and generically.
> 
> The Serde ecosystem consists of data structures that know how to serialize and deserialize themselves along with data formats that know how to serialize and deserialize other things. Serde provides the layer by which these two groups interact with each other, allowing any supported data structure to be serialized and deserialized using any supported data format.

In other words,

> data structures that know how to serialize and deserialize themselves

^^ These are your built-in/intrinsic data types

> along with data formats that know how to serialize and deserialize other things

^^ these "data formats" are your formatters.  the "other things" are complex data types built out of intrinsics.

> Serde provides the layer by which these two groups interact with each other, allowing any supported data structure to be serialized and deserialized using any supported data format.

^^ This layer is what you use as a magic trick to allow serializing "any supported data structure" (complex types built out of intrinsics).

**Reply by bradwilson** [reply]
I haven't investigated that framework, but the problem isn't things that know how to serialize and deserialize themselves, it's putting a generic "what type should I create?" into the serialization stream where the security issue comes from. (By which I mean, not a fixed lookup table, but the ability to arbitrarily create instances of any type in the type system.)

**Reply by jzabroski** [reply]
Right, and the magic trick is that MyTests.dll is the authoritative reference for both sides of what extends the fixed lookup table.

This is the part I'm not 100% is secure, but I think it is. Whether it is something that works with Visual Studio architecture is another question. But... It's not much different from how a Test class has a reference to ITestOutputHelper.  The test defines what object-capability security people call a Powerbox of capabilities the Test class instance has access to.

An attacker would need to somehow replace the dll to confuse the system.

**Reply by jzabroski** [reply]
Hi @bradwilson ,
I've been discussing this serialization issue with a friend.  I think, practically, xunit.net needs a threat model before making architecture changes.  Whether to discuss a threat model publicly is up to you, but probably also requires discussion with VS Engineering and JetBrains.  I also spent some time studying how JetBrains hooks into your xunit.testrunner, but it's time consuming and I've only made some progress.

**Reply by markusschaber** [reply]
> I personally think that the intrinsic types are enough. I can't remember ever looking or needing more than that. So from my POV, your first alternative would be fine.

We currently have lots of Uri and some GUIDs in our theory data, they're currently not in the list of supported data types.

**Comment by jzabroski** [comment]
I prefer ability to run individual theory rows individually. But - is that limitation due to Visual Studio Test Explorer being lame? I use ReSharper so I don't know if this would affect me.

What problem are you solving by removing this?

EDIT: I see the problem.

> Binary serialization has been [deprecated because of security concerns](https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide) and will be outright removed from future versions of .NET. Other related serializers (SoapFormatter, LosFormatter, NetDataContractSerializer, ObjectStateFormatter) suffer from the same concerns.

A lot easier to read discussions on Desktop than mobile phone...

**Comment by SimonCropp** [comment]
this would break a few things for me. But all of those can be worked around. And IMO removing ISerializable support is a good idea to remove the associated complexity

**Reply by bradwilson** [reply]
I want to say complexity will go down, but I'm not sure that's necessarily true if I re-introduce `IXunitSerializable`. Some of the complexity will shift to supporting that (effectively the same as the previous support for that, except that now it has stronger type limitations to prevent polymorphic object creation).

**Reply by bradwilson** [reply]
That is, assuming I go with strategy 2 and not strategy 1.

**Reply by markusschaber** [reply]
Instead of introducing a new interface which is to be implemented by the data types, provide a way to reference a serialization mechanism, for example via an `[Attribute]` at the TheoryData member, or the class or assembly level. This has the advantage that existing datatypes (which sometimes cannot be modified) can still be made serializable for the test theories.

**Comment by markm77** [comment]
It is extremely important for me to be able to run theory rows individually. Currently I am using `IXunitSerializable` and split my theory data into the following two classes which allows for custom compact test runner labelling via ToString():

`BankTestData1`: https://github.com/finlabsuk/open-banking-connector/blob/02f1b903321d07e59e60d3cca60a95a228902269/test/OpenBanking.Library.Connector.BankTests/BankTests/BankTestData1.cs
`BankTestData2`: https://github.com/finlabsuk/open-banking-connector/blob/02f1b903321d07e59e60d3cca60a95a228902269/test/OpenBanking.Library.Connector.BankTests/BankTests/BankTestData2.cs

You can see in the above classes I am using custom serialisation for some properties.

My tests do use the same types returned by the TheoryData. For example: https://github.com/finlabsuk/open-banking-connector/blob/02f1b903321d07e59e60d3cca60a95a228902269/test/OpenBanking.Library.Connector.BankTests/BankTests/GenericHostAppTests.cs#L24-L47

Would be very disappointing to lose either ability to run theory rows individually or even ability to customise labelling in test runner by grouping in classes as above....

**Reply by jzabroski** [reply]
@markm77 Something neat is coming to C# 11 that could be a game changing feature for TheoryData: [Generic Attributes](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/generic-attributes.md). Note, it's in the C# 10.0 proposal folder, but I believe it's been pushed to C# 11 and is in Preview mode now.

**Comment by bradwilson** [comment]
I've pushed up into `dev` a removal of `ISerializable` and re-introduction of `IXunitSerializable`: https://github.com/xunit/xunit/commit/87e058d4892c3a1ab97573f64d838767ad5a6b23

The implementation supports:

- Intrinsic types: `String`, `Char`, `Byte`, `SByte`, `Int16`, `UInt16`, `Int32`, `UInt32`, `Int64`, `UInt64`, `Single`, `Double`, `Decimal`, `Boolean`, `DateTime`, `DateTimeOffset`, `Type`
- Enum values
- Nullable versions of the value-type intrinsics and enum values
- Custom types implementing `IXunitSerializable`
- Arrays of supported types (f.e., `String[]`)
- Method parameter lists (aka `Object[]` that contains serializable values)
- Traits dictionaries (aka `Dictionary<string, List<string>>`)

Anything not on this list won't be supported; in particular, the generic collection class like `List<>` are not supported (with the exception of trait dictionaries, which have both hard-coded types as well as hard-coded comparers). I believe in practice, this should allow any test from v2 currently shows up as individually runnable in Visual Studio Test Explorer today should continue to show up as runnable in v3.

Just as important, since types are still embedded (albeit safely now, see discussion below), this will _not_ impact the parameter types you can use; you are still able to use things like `object` or generics to represent polymorphic theory data.

Based on the discussion here, this implementation does a safe(r) form of type embedding in the serialization: rather than always embedding type names it uses a type index, and we only embed full types names for enums and `IXunitSerializable`. I consider this to be safe(r) because you can no longer create arbitrary types; besides the intrinsics (which should all be safe), enum values are limited to know enum types (which should all be safe) and `IXunitSerializable` values are limited to only things which implement `IXunitSerializable`, which means they're explicitly designed to be serialized and deserialized only in the context of the testing framework.

Hopefully this will address the concerns with the loss of flexibility by rolling back the `ISerializable` support while also address the arbitrary type creation concerns of the .NET team.

I haven't merged this into `main` yet, so there is no NuGet-based build, but you wish to review the source and provide feedback, that would be most welcome. The best place to start to understand what the code does is, as usual, the tests; in particular, [`SerializationHelperTests.cs`](https://github.com/xunit/xunit/blob/87e058d4892c3a1ab97573f64d838767ad5a6b23/src/xunit.v3.common.tests/Serialization/SerializationHelperTests.cs) and [`XunitSerializationInfoTests.cs`](https://github.com/xunit/xunit/blob/87e058d4892c3a1ab97573f64d838767ad5a6b23/src/xunit.v3.common.tests/Serialization/XunitSerializationInfoTests.cs) (the latter is used for serializing `IXunitSerializable` objects as well as [arrays](https://github.com/xunit/xunit/blob/87e058d4892c3a1ab97573f64d838767ad5a6b23/src/xunit.v3.common/Serialization/SerializationHelper.cs#L410-L416)).

**Reply by SimonCropp** [reply]
DateOnly and TimeOnly?

**Reply by bradwilson** [reply]
Currently beyond the minimum requirements (not supported in .NET Standard 2.0). If it's important enough, it could be done via reflection, perhaps.

**Reply by jzabroski** [reply]
I guess that's fine, but...

> while also address the arbitrary type creation concerns of the .NET team.

I still don't think we know what those concerns were. What is xunit.net's threat model given its an IDE plug-in (xunit.testrunner)?

**Reply by bradwilson** [reply]
@jzabroski The v3 behavior will be essentially identical to the v2 behavior. Given that essentially all forms of type-embedded serialization built into .NET have been deprecated and may be removed in the future, it's a bad future-looking decision for us to rely on them.

**Reply by jzabroski** [reply]
OK, I can get on board with that reasoning, but the previous comment made it sound like you had specific advice given to you, and that it would benefit the project to share it.

Overall, thank you for this discussion, as it made me think of some stuff I haven't had to think about in years (object capability security, taming, application threat modeling, etc.)

**Reply by bradwilson** [reply]
FYI, I merged this into `main` today.

**Reply by markusschaber** [reply]
>     * Intrinsic types: `String`, `Char`, `Byte`, `SByte`, `Int16`, `UInt16`, `Int32`, `UInt32`, `Int64`, `UInt64`, `Single`, `Double`, `Decimal`, `Boolean`, `DateTime`, `DateTimeOffset`, `Type`
> 
> [...] I believe in practice, this should allow any test from v2 currently shows up as individually runnable in Visual Studio Test Explorer today should continue to show up as runnable in v3.

Would you consider adding `System.Guid` and `System.Uri` to the list of intrinsics? We have quite a lot of tests using Uris, and a few using Guids in the theory data.

**Reply by bradwilson** [reply]
> Would you consider adding System.Guid and System.Uri to the list of intrinsics?

Yes. Open an issue.

**Reply by markusschaber** [reply]
The issue is: https://github.com/xunit/xunit/issues/3038 - if I missed anything, just tell me, and I'll update it. Thanks!

**Comment by alexTr3** [comment]
I'm on 3.2.1 mtp v2 and have this bug.. wtf....

**Reply by bradwilson** [reply]
What bug, exactly? Maybe you should open a new issue with a repro project.
